File: Practical-1/Practical1_3.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Practical1_3 {
    private static final int TABLE_SIZE = 10;  // Example hash table size
    private static int[] hashTableLinear = new int[TABLE_SIZE];
    private static int[] hashTableQuadratic = new int[TABLE_SIZE];

    public static void main(String[] args) {
        // Initialize the hash tables
        for (int i = 0; i < TABLE_SIZE; i++) {
            hashTableLinear[i] = -1;  // -1 means empty
            hashTableQuadratic[i] = -1;
        }

        try {
            // Step 1: Read numbers from a file
            File file = new File("/Users/bhavik/Downloads/numbers.txt");
            Scanner sc = new Scanner(file);

            int linearCollisions = 0;
            int quadraticCollisions = 0;

            while (sc.hasNextInt()) {
                int num = sc.nextInt();
                
                // Step 2: Insert using linear probing
                if (!insertLinear(num)) {
                    linearCollisions++;
                }
                
                // Step 3: Insert using quadratic probing
                if (!insertQuadratic(num)) {
                    quadraticCollisions++;
                }
            }

            sc.close();
            
            // Output the number of collisions
            System.out.println("Linear Probing Collisions: " + linearCollisions);
            System.out.println("Quadratic Probing Collisions: " + quadraticCollisions);

        } catch (FileNotFoundException e) {
            System.out.println("File not found!");
        }
    }

    // Insert using linear probing
    private static boolean insertLinear(int num) {
        int hash = num % TABLE_SIZE;
        int originalHash = hash;
        boolean collision = false;

        while (hashTableLinear[hash] != -1) {
            collision = true;
            hash = (hash + 1) % TABLE_SIZE;  // Linear probing
            if (hash == originalHash) {
                return false;  // Table is full
            }
        }

        hashTableLinear[hash] = num;
        return collision;
    }

    // Insert using quadratic probing
    private static boolean insertQuadratic(int num) {
        int hash = num % TABLE_SIZE;
        int originalHash = hash;
        boolean collision = false;
        int i = 1;

        while (hashTableQuadratic[hash] != -1) {
            collision = true;
            hash = (originalHash + i * i) % TABLE_SIZE;  // Quadratic probing
            i++;
            if (i > TABLE_SIZE || hash == originalHash) {
                return false;  // Table is full or no slot available
            }
        }

        hashTableQuadratic[hash] = num;
        return collision;
    }
}

File: Practical-1/Practical1-3.py
#  insert and retrieve numbers from a file and then output the number of collisions for:

# Linear Probing
# Quadratic Probing

TABLE_SIZE = 10

# Initialize hash tables
hash_table_linear = [-1] * TABLE_SIZE
hash_table_quadratic = [-1] * TABLE_SIZE


def insert_linear(num):
    """Insert using linear probing and return True if collision occurs."""
    hash_val = num % TABLE_SIZE
    original_hash = hash_val
    collision = False

    while hash_table_linear[hash_val] != -1:
        collision = True
        hash_val = (hash_val + 1) % TABLE_SIZE
        if hash_val == original_hash:
            return collision  # Table is full, collision occurred

    hash_table_linear[hash_val] = num
    return collision


def insert_quadratic(num):
    """Insert using quadratic probing and return True if collision occurs."""
    hash_val = num % TABLE_SIZE
    original_hash = hash_val
    collision = False
    i = 1

    while hash_table_quadratic[hash_val] != -1:
        collision = True
        hash_val = (original_hash + i * i) % TABLE_SIZE
        i += 1
        if i > TABLE_SIZE or hash_val == original_hash:
            return collision  # Table is full, collision occurred

    hash_table_quadratic[hash_val] = num
    return collision


def main():
    # Initialize collision counts
    linear_collisions = 0
    quadratic_collisions = 0

    try:
        # Step 1: Read numbers from the file
        with open('/Users/bhavik/Downloads/numbers.txt', 'r') as file:
            for line in file:
                num = int(line.strip())
                
                # Step 2: Insert into both hash tables and count collisions
                if insert_linear(num):
                    linear_collisions += 1
                if insert_quadratic(num):
                    quadratic_collisions += 1

        # Step 3: Output the number of collisions
        print("Linear Probing Collisions:", linear_collisions)
        print("Quadratic Probing Collisions:", quadratic_collisions)

    except FileNotFoundError:
        print("File not found!")


if __name__ == "__main__":
    main()
